enhancements/proxy/global-cluster-egress-proxy.md

```diff
@@ -0,0 +1,285 @@
+environment variables injected (HTTP_PROXY, HTTPS_PROXY, NO_PROXY) by the cluster version operator.
+
+Critical touch points for the administrator:
+
+* edits cluster scoped proxy config resource (spec fields)
+* provides a configmap of additional CAs
+
+Critical touch points for proxy configuration consumers:
+
+* Operator consumes status fields from cluster scoped proxy config resource and updates its operand accordingly
+* *May* consume the "accepted CAs" configmap (openshift-config-managed/trusted-ca-bundle) to get CAs
+* *Should* create their own configmap with the `config.openshift.io/inject-trusted-cabundle: "true"` label and
+consume the CA bundle from there.
+* Operator deployment may request proxy environment injection via the `config.openshift.io/inject-proxy: <container-name>`
+annotation since operators cannot control their own environment variables, but the oprator is responsible for mounting
+a configmap to pick up the CAs if it needs them.
+
+Automatic no proxy behavior:
+
+The following domains/hosts are automatically added to the no proxy configuration:
+* TODO
+
+Known limitations/future enhancements:
+
+* The no proxy entries that are automatically added do not include the cluster domain generically, so
+requests that go to external routes will likely go through the proxy.  This may be undesirable, though
+it should not break as long as the proxy is able to call back to the cluster.  The administrator can add 
+additional no proxy entries.  It was deemed unacceptable to automatically noproxy the entire cluster domain 
+because the front end load balancing/routing for the cluster could reside outside the clusters network.
+* Not all cloud service apis are added to the no proxy list.  This means some cloud api requests may
+go through the proxy, which again is undesirable though it should not break things as long as the proxy
+is functional.  The administrator can add additional no proxy entries.
+* No validation endpoints can be provided at install time.  This is because there is no point in validating
+the proxy configuration during install.  If it is incorrect, the install will fail.  However it also 
+puts administrators at risk of never providing validation endpoints to their proxy configuration, which
+means they can update their proxy configuration to something that does not work, in the future, if they do
+not add validation endpoints on day 2.  We should consider enhancing the install config to allow specification
+of validation endpoints so this step is not forgotten by administrators.
+* Currently the no proxy value in the proxy config is append-only.  There is no way for an administrator to
+remove one of the no proxy domains that we add automatically.  This means we must be extremely cautious to
+not add noproxy domains that might need to be proxied.
+* Long term the additional CA bundle should instead be the *only* source of CAs for components.  Today the additional
```

Comment:
In previous discussions, I thought the consensus was to support both ca use cases 1) append ca's 2) replace ca's?

Reply:
that's not my recollection...  as i recall we said if someone wants to provide a bundle, then they are providing the entire bundle.  But we can negotiate that if/when we get to the point of implementing such a capability.  Certainly the delta as to whether the proxy bundle operator logic takes what is given, or appends the system CAs(as it does today) is pretty small from a code perspective.

Reply:
@bparees [this](https://coreos.slack.com/archives/CLJSH16J0/p1564587559136400) is the slack conversation I'm referencing. My takeaway from the conversation is 90% of the user-base wants to add a trusted ca to the bundle, while 10% (paranoid customers) want to replace the entire bundle.

/cc @abhinavdahiya @smarterclayton

Reply:
@danehans you're recollection is correct (thanks for the reference).  I will update this to make it clear the long term improvement is to make it configurable whether the user-provided-bundle gets augmented w/ the system trusts or not.